### PR TITLE
add integration test for storm

### PR DESCRIPTION
**Pull request by manuzhang**
[Review on Reviewable](https://reviewable.io/reviews/gearpump/gearpump/1583)

### CODE REVIEW

File: integrationtest/docker/jdk8/start.sh [hunk 1]
```diff
@@ -36,6 +36,9 @@ case "$COMMAND" in
     JAVA_OPTS="$JAVA_OPTS -Dgearpump.hostname=$(hostname)"
     sh "$SUT_HOME"/bin/"$COMMAND" "$@"
     ;;
+  drpc)
```
Comment (stanleyxu2005) [inline integrationtest/docker/jdk8/start.sh]: storm-drpc

File: integrationtest/docker/jdk8/start.sh [hunk 2]
```diff
@@ -36,6 +36,9 @@ case "$COMMAND" in
     JAVA_OPTS="$JAVA_OPTS -Dgearpump.hostname=$(hostname)"
     sh "$SUT_HOME"/bin/"$COMMAND" "$@"
     ;;
+  storm-drpc)
+    java -server -Xmx768m -cp "$SUT_HOME"/lib/*:"$SUT_HOME"/lib/storm/* backtype.storm.daemon.drpc
```
Comment (stanleyxu2005) [inline integrationtest/docker/jdk8/start.sh]: Do we need to restrict `-Xmx768m`?

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala [hunk 3]
```diff
@@ -19,52 +19,72 @@ package io.gearpump.integrationtest.checklist
...
     "support DRPC" in {
-
-    }
-
-    "support at-least-once semantics with Kafka" in {
-
+      stormClient.submitStormApp(mainClass = "storm.starter.ReachTopology",
+        args = "reach")
+      val drpcClient = stormClient.getDRPCClient
+      drpcClient.execute("reach", "foo.com/blog/1") shouldBe "16"
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala]: The data is not feed explicitly before you submit a storm app. Where these results 16, 14, 0 come from? At least update the test case description. Change "Support DRPC" to something more specific about what is being tested.

BTW: I suggest use `shouldEqual` for values except `shouldBe true` or `shouldBe false`.

File: integrationtest/src/main/scala/io/gearpump/integrationtest/Shell.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gearpump.integrationtest
+
+import org.apache.log4j.Logger
+import sys.process._
+
+object Shell {
+  private val LOG = Logger.getLogger(getClass)
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/Shell.scala]: Please always put one line before and after class's `{}`.
```
object Shell {

   // something
   // .....

   // leave a line blank before last }

}

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/CommandLineClient.scala [hunk 5]
```diff
@@ -73,5 +69,4 @@ class CommandLineClient(host: String) {
   private def execAndCaptureOutput(command: String): String = {
     Docker.execAndCaptureOutput(host, s"/opt/start $command")
   }
-
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/CommandLineClient.scala]: keep the empty line

File: integrationtest/docker/jdk8/start.sh (hunk 2 above)
Comment (manuzhang) [inline integrationtest/docker/jdk8/start.sh]: consistent with storm command

File: integrationtest/src/main/scala/io/gearpump/integrationtest/storm/StormClient.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  private val STORM_HOST = "storm0"
+  private val STORM_CMD = "/opt/start storm"
+  private val STORM_STARTER_JAR = "/opt/gearpump/lib/storm/storm-starter-0.9.5.jar"
+  private val STORM_DRPC = "storm-drpc"
+  private val IMAGE = "stanleyxu2005/gpct-jdk8"
+  private val CONFIG_FILE = "storm.yaml"
+  private var drpcServerIp: String = null
+
+  def start(): Unit = {
+    Docker.createAndStartContainer(STORM_HOST,
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/storm/StormClient.scala]: What is the propose of creating a storm client? You can expose more ports in MiniCluster's constructor. Do you only want to obtain special ip when cluster is getting started?

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala [hunk 7]
```diff
@@ -19,52 +19,72 @@ package io.gearpump.integrationtest.checklist
...
 import io.gearpump.integrationtest.minicluster.Util
+import io.gearpump.integrationtest.storm.StormClient
 
 /**
  * The test spec checks the compatibility of running Storm applications
  */
 class StormCompatibilitySpec extends TestSpecBase {
 
-  private val STORM_STARTER_JAR = "/opt/gearpump/lib/storm/storm-starter-0.9.5.jar"
+  private val stormClient = new StormClient(cluster, commandLineClient)
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala]: It is already to late to override a cluster in this way. Cluster was started once in StandaloneModeSpec, before `beforeAll` here.

File: integrationtest/docker/jdk8/Dockerfile [hunk 8]
```diff
@@ -12,4 +12,6 @@ WORKDIR $SUT_HOME
 ADD start.sh /opt/start
 RUN chmod +x /opt/start
 
+COPY storm.yaml SUT_HOME
```
Comment (stanleyxu2005) [inline integrationtest/docker/jdk8/Dockerfile]: No need to add a real file to Docker image. You have already create one on the fly using `echo content > file`

File: integrationtest/docker/jdk8/start.sh [hunk 9]
```diff
@@ -36,12 +36,19 @@ case "$COMMAND" in
     JAVA_OPTS="$JAVA_OPTS -Dgearpump.hostname=$(hostname)"
     sh "$SUT_HOME"/bin/"$COMMAND" "$@"
     ;;
+  storm-drpc)
+    STORM_CONFIG="$SUT_HOME"/storm.yaml
+    echo "drpc.servers:" > "$STORM_CONFIG"
+    echo "      "-" "\"`ip route | awk '/default/ {print $3}'`\" >> "$STORM_CONFIG"
```
Comment (stanleyxu2005) [inline integrationtest/docker/jdk8/start.sh]: Maybe the command is suitable for Mac OS, but it returns the real IP of Docker host on Linux. I assume we want the IP address on `docker0` network.
```
qxu@ubuntu:~$ ip route
default via 192.168.31.1 dev eth0 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 
192.168.31.0/24 dev eth0  proto kernel  scope link  src 192.168.31.200 
qxu@ubuntu:~$ ip route | awk '/default/ {print $3}'
192.168.31.1
```

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala [hunk 10]
```diff
@@ -19,52 +19,72 @@ package io.gearpump.integrationtest.checklist
...
+  }
+
+  "Storm over Gearpump" should {
+    "support basic topologies" in {
       // exercise
-      val appId = commandLineClient.submitStormApp(STORM_STARTER_JAR,
-        "storm.starter.ExclamationTopology exclamation")
+      val appId = stormClient.submitStormApp(
+        mainClass = "storm.starter.ExclamationTopology", args = "exclamation")
       Thread.sleep(5000)
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala]: Is the sleep really needed?

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala [hunk 11]
```diff
@@ -19,52 +19,72 @@ package io.gearpump.integrationtest.checklist
...
+      drpcClient.execute("reach", "notaurl.com") shouldBe "0"
     }
 
     "support tick tuple" in {
       // exercise
-      val appId = commandLineClient.submitStormApp(STORM_STARTER_JAR,
-        "storm.starter.RollingTopWords slidingWindowCounts remote")
+      val appId = stormClient.submitStormApp(mainClass = "storm.starter.RollingTopWords",
+        args = "slidingWindowCounts remote")
       Thread.sleep(5000)
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/StormCompatibilitySpec.scala]: Is the sleep really needed

**Comment by stanleyxu2005** [comment]
Please minimize your PR with one test case and StormClient changes. Other test cases can be added in separate PR. Let's keep the review process effective.

File: integrationtest/docker/jdk8/start.sh (hunk 9 above)
Comment (manuzhang) [inline integrationtest/docker/jdk8/start.sh]: ip address of docker0 is the docker host on linux, which is how the host connect ports exposed by containers inside docker0

**Comment by manuzhang** [comment]
updated

**Comment by stanleyxu2005** [comment]
+1

**Pull request merged**